nav2_smac_planner/src/node_hybrid.cpp

```diff
@@ -492,6 +515,126 @@ float NodeHybrid::getObstacleHeuristic(
   return 2.0 * obstacle_heuristic_lookup_table[start_index];
 }
 
+inline float distanceHeuristic2D(
+  const int idx, const int size_x, const int goal_x,
+  const int goal_y)
+{
+  int x = std::abs(idx % size_x - goal_x);
+  int y = std::abs(idx / size_x - goal_y);
+  float & table_value = NodeHybrid::astar_2d_h_table[y * size_x + x];
+  return table_value < 0.0f ? table_value = std::hypotf(x, y) : table_value;
+}
+
+float NodeHybrid::getObstacleHeuristicAdmissible(
+  const Coordinates & node_coords,
+  const Coordinates & goal_coords,
+  const double & cost_penalty)
+{
+  int size_x = sampled_costmap->getSizeInCellsX();
+  // Divided by 2 due to downsampled costmap.
+  const int start_y = floor(node_coords.y / 2.0);
+  const int start_x = floor(node_coords.x / 2.0);
+  const int start_index = start_y * size_x + start_x;
+  const float & starting_cost = obstacle_heuristic_lookup_table[start_index];
+  if (starting_cost > 0.0f) {
+    // costs are doubled due to downsampling
+    return 2.0 * starting_cost;
+  }
+
+  const int size_x_int = static_cast<int>(size_x);
+  const int size_y = sampled_costmap->getSizeInCellsY();
+
+  const std::vector<int> neighborhood = {1, -1,  // left right
+    size_x_int, -size_x_int,  // up down
+    size_x_int + 1, size_x_int - 1,  // upper diagonals
+    -size_x_int + 1, -size_x_int - 1};  // lower diagonals
+
+  const int goal_x = floor(goal_coords.x / 2.0);
+  const int goal_y = floor(goal_coords.y / 2.0);
+  const int goal_index = goal_y * size_x + goal_x;
+  if (astar_2d_first_run) {
```

Comment:
I believe (or perhaps not start?) that you have all of the information required to do the `astar_2d_queue.emplace_back(...)` in `resetObstacleHeuristic`. That would let you remove the `astar_2d_first_run` flag and streamline this function. That's what the original implementation did, e.g.

```
obstacle_heuristic_queue.emplace(
      ceil(goal_y / 2.0) * sampled_costmap->getSizeInCellsX() + ceil(goal_x / 2.0));
```

after it resized all of the data structures and reset them. That would keep the 2 implementations at least relatively analog.

Reply:
This was done just because of the start position missing in `resetObstacleHeuristic`. No problem to change the method signature and put it there instead.

Reply:
If we're going to remove the current heuristic implementation, there's no reason to change the signature since there wouldn't be 2 implementations to try to keep analog. We can keep it this way right now. I'll recircle back to this comment after we benchmark performance and come to a final decision about whether to remove it.